[binary addons] add bootstrapping logic to download addon build definitions from separate repository

Montellese opened the pull request:
This is the next step towards being able to build specific binary addons from a binary addons repository and uploading them to the repository on our mirros.
- It introduces a new variable `ADDONS_DEFINITION_DIR` which specifies where the binary addon build definitions are located (which is currently `project/cmake/addons/addons` which is also the default vaue of the new variable). This allows us to place these directories and files anywhere we want as long as we properly set the value of the variable.
- It introduces a new cmake based "bootstrap" buildsystem which can be executed before building the binary addons. Its purpose is to download binary addon build definitions from another git repository (see e.g. https://github.com/Montellese/audiodecoder-addons) which can then be used to build the binary addons using our existing buildsystem. It expects the following variables to be set: `CMAKE_INSTALL_PREFIX` should be set to the place where the binary addon buildsystem expects the binary addon build definitions. `ADDONS_TO_BUILD` can limit the addons to actually build. `ADDON_TYPE` must be the type of the addons to be bootstrapped (e.g. `audiodecoder`). `ADDON_DEFINITIONS_REPOSITORY` is the URL of the git repository containing the binary addon build definitions to be downloaded. `ADDON_DEFINITIONS_REVISION` can be used to specify a revision to be downloaded from the given git repository.
- For WIN32 there's an incomplete `bootstrap-addons.bat` which takes most of the previously mentioned variable values as arguments and executes the bootstrap buildsystem.

So there would be an additional build step in the binary addons build process which would need to be executed before being able to build any binary addons. The general idea comes from @notspiff (big surprise ;-)) and has a few advantages:
- binary addon build definitions can be removed from Kodi's main git repository
- we can create and maintain separate git repositories for separate binary addon types (audioencoders, audiodecoders, pvr addons, ...)
  Obviously there's also a downside which needs to be considered and discussed:
- the common dependencies `platform` and `kodi-platform` are still defined in Kodi's git repository and there's no explicit relationship anymore between the binary addons versions and the dependencies' versions. So building binary addons for an older release would require to checkout an older version of Kodi's git repository and an older version of the binary addons repositories.

A few TODO's that came to mind (there are probably more):
- integrate the bootstrap buildsystem into the buildsteps of all platforms
- fix and document `bootstrap-addons.bat` for WIN32
- ~~figure out whether we should provide default definitions for the different binary addons repositories which are used if no explicit repository is given (to still be able to test building all binary addons)~~
- ~~move all binary addons build definitions into separate git repositories per binary addon type (keep the ones that should be shipped with Kodi by default)~~
- adjust jenkins build jobs to support specifying the binary addons to be built

Montellese committed:
[binary addons] add ADDONS_DEFINITION_DIR to be able to specify the directory containing the definitions for the addons to be built

Montellese committed:
[binary addons] replace all addons with a binary addons definition repository

Montellese committed:
[binary addons] provide a bootstrap buildsystem to download addon definitions

Montellese committed:
[win32] provide bootstrap-addons.bat to execute binary addons bootstrapping

wsnipex commented on project/cmake/addons/bootstrap/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+
+# make sure that the addon type has been specified
+if(NOT ADDON_TYPE)
+  message(FATAL_ERROR "No addon type specified")
```
this seems unintuitive. ADDONS_TO_BUILD is already used in the normal addon project and takes a list of arbitrary addons, no need to match them to specific types. I think we should retain this behavior here as well.

notspiff replied:
The value is used for a directory name only.
10. aug. 2015 20:45 skrev "Wolfgang Schupp" <notifications@github.com>:

> In project/cmake/addons/bootstrap/CMakeLists.txt
> <https://github.com/xbmc/xbmc/pull/7774#discussion_r36668892>:
>
> > +else()
> > +  file(TO_CMAKE_PATH "${BUILD_DIR}" BUILD_DIR)
> > +endif()
> > +get_filename_component(BUILD_DIR "${BUILD_DIR}" ABSOLUTE)
> > +
> > +# figure out which addons to bootstrap (defaults to all)
> > +if(NOT ADDONS_TO_BUILD)
> > +  set(ADDONS_TO_BUILD "all")
> > +  message(STATUS "Bootstrapping all addons")
> > +else()
> > +  message(STATUS "Bootstrapping following addons: ${ADDONS_TO_BUILD}")
> > +endif()
> > +
> > +# make sure that the addon type has been specified
> > +if(NOT ADDON_TYPE)
> > +  message(FATAL_ERROR "No addon type specified")
>
> this seems unintuitive. ADDONS_TO_BUILD is already used in the normal
> addon project and takes a list of arbitrary addons, no need to match them
> to specific types. I think we should retain this behavior here as well.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/7774/files#r36668892>.
>

Montellese replied:
The only reason I added this is that we probably want to have multiple repositories for different binary addons types so you have a repository containing all audioencoders, one containing all audiodecoders and so on. And you need to checkout these repositories into separate directories because otherwise you end up with one bootstrapping overwriting another. Therefore you need to provide a name (I went for the type) for the binary addons repository that you want to bootstrap. Obviously we can rename the variable if `ADDON_TYPE` is confusing.

wsnipex replied:
how about using the same txt file based definition as for the addons?

wsnipex commented on project/cmake/addons/bootstrap/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+
+# make sure that the addon definitions repository has been specified
+if(NOT ADDON_DEFINITIONS_REPOSITORY)
+  message(FATAL_ERROR "No addons definition repository specified")
```
imho we must provide a default repo that holds all addons.

wsnipex replied:
as an example, see how I do it in my ubuntu packaging script:
https://github.com/wsnipex/xbmc-packaging/blob/binary-addons/build-xbmc-addons.sh#L42-L88

Montellese replied:
Yeah I'm not sure yet how to do that because there will be multiple repositories (i.e. not one repositories with all binary addon definitions) and it feels wrong to hardcode them in here. We could do something similar and allow one `.txt` file per repository and if no repository is provided we read those. But that would require some refactoring because right now this buildsystem is only capable of handling one repository at a time so it would have to be called multiple times.

wsnipex replied:
Whats the advantage of having multiple repos? 
I'd much prefer a single meta repo that holds all addons, if possible separated into addon classes/types. Imho this is easier to handle and to not lose overview.

notspiff replied:
Because third parties.
11. aug. 2015 16:06 skrev "Wolfgang Schupp" <notifications@github.com>:

> In project/cmake/addons/bootstrap/CMakeLists.txt
> <https://github.com/xbmc/xbmc/pull/7774#discussion_r36748021>:
>
> > +# figure out which addons to bootstrap (defaults to all)
> > +if(NOT ADDONS_TO_BUILD)
> > +  set(ADDONS_TO_BUILD "all")
> > +  message(STATUS "Bootstrapping all addons")
> > +else()
> > +  message(STATUS "Bootstrapping following addons: ${ADDONS_TO_BUILD}")
> > +endif()
> > +
> > +# make sure that the addon type has been specified
> > +if(NOT ADDON_TYPE)
> > +  message(FATAL_ERROR "No addon type specified")
> > +endif()
> > +
> > +# make sure that the addon definitions repository has been specified
> > +if(NOT ADDON_DEFINITIONS_REPOSITORY)
> > +  message(FATAL_ERROR "No addons definition repository specified")
>
> Whats the advantage of having multiple repos?
> I'd much prefer a single meta repo that holds all addons, if possible
> separated into addon classes/types. Imho this is easier to handle and to
> not lose overview.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/7774/files#r36748021>.
>

wsnipex replied:
not following. Why would 3rd parties be blocked by our meta repo being a single one. I'm not saying other repos should not be possible.
Can someone outline this, so we are all on the same page?

akva2 replied:
right. as long as multiple repos are supported idk.

Montellese replied:
IMO we need to support bootstrapping being executed multiple times for different meta repositories (or adjust it to be able to handle multiple meta repositories which are defined somewhere) independent of whether we ourselves will use it. And that means that we need to choose a name to be used in `externalproject_add()`. I'm fine with adding default values for `ADDON_TYPE` (e.g. `all`) and `ADDON_DEFINITIONS_REPOSITORY` but if someone wants to bootstrap multiple meta repositories and then build addons from both meta repositories we need to be able to choose different names to be used in `externalproject_add()`.

The reason why I would prefer multiple meta repositories is that it would make it easier to e.g. only build all PVR addons (if something has changed in the PVR API) without having to manually define all the PVR addons that need to be built. You'd just call this bootstrap buildsystem with the meta repo containing all PVR addon definitions and you'd be good to go.

Montellese commented:
@wsnipex: Refactored this a bit to
* support multiple meta repositories
* (multiple) meta repository definition files
* possibility to specify a single meta repositiory to be bootstrapped (must have a definition file) with the possibility to specify a specific revision (overruling the one from the definition file)
* possibility to specify the addons to be bootstrapped

Does that sound better for you?

wsnipex commented:
@Montellese yes, that looks much bettter, thanks.
Only 1 question: Can we have the default repos bootstrapped automatically, if no other repo was enabled?
That would retain old behavior for e.g. unified deps.

Montellese commented:
@wsnipex: I guess it depends on what you mean by "bootstrapped automatically". You need to execute the bootstrap cmake buildsystem or you'll end up without binary addons. But if you just call cmake on it with the necessary parameters (without a specific repository or specific addons to build) it will bootstrap based on all the repository definition files it can find (similar to how the addons buildsystem works).

wsnipex commented:
I mean, run bootstrap from the addons CMakeLists if no addons were found / bootstrap was not done yet.
You could create a "bootstrap.done" file in bootstrap and test that from the binary addons build sys.
I'd like to avoid introducing another step that needs to be executed by users and scripts for the default case.

Montellese commented:
I'll have to check because the entire bootstrap buildsystem would have to be executed (not just generated) before starting to process any addons in the addons buildsystem.

Montellese committed:
[binary addons] automatically bootstrap all default repositories and addons if no addon definitions are available

Montellese commented:
@wsnipex: Done, see 5538887f2093fbbd463c7d114627f528fe219a60 which checks if there are any addon definitions available and if not automatically generates and executes the bootstrap buildsystem therefore bootstrapping all default meta repositories and their addons.

Either way, IMO we still need the possibility to specify a specific meta repository (including revision) and addon to be built for integration into jenkins so that we can trigger jenkins builds for specific repositories and even specific addons. But that can also come in a next step.

wsnipex commented:
nice, thanks!

Montellese commented:
@MartijnKaijser: what would be the preferred way for binary addon repositories? One repository containing all binary addon definitions (independent of their type (PVR, audio encoders, ...)) or one repository for every binary addon type i.e. one for PVR addons, one for audio encoders and so on?

MartijnKaijser commented:
one binary repo (per platform i presume) would be best/easiest imo. Will also need some way for packaging for creating the addons.xml list. Either open the zip and read the addon.xml or copy the addon.xml next to the zip file

notspiff commented:
i wrote a python thingie for that which i sent to montellese half a year ago or something.

Montellese commented:
@MartijnKaijser: Sorry I didn't express that question very well. I didn't mean the actual addon repositories. I was talking about the git repositories containing the definitions for the addons to be built, see e.g. https://github.com/Montellese/audiodecoder-addons which contains the definitions for all audiodecoder-addons. The resulting addon ZIPs would then be copied into the actual addon repositories which can be used within Kodi.

MartijnKaijser commented:
ah, well don't know. We do have them split for plugins,scripts,skins,resource which just makes it a bit more "clean". Though there are a lot more of them and cloning on giant big git would be a pain.
Having them split doesn't cause real pain for handling them so i'd go for split per type. oth you would have to clone each repo for making changes/pushing instead of just a big one. No preference from my side

Montellese commented:
Let's see if this also works on other platforms. jenkins build this please

Montellese commented:
This might be the first time that a change in the CMake buildsystem compiled on all platform :-)

wsnipex commented:
lol indeed :+1:

akva2 commented:
you should probably use branches in the repo gits, not master.

Montellese commented:
Yeah, this is just for testing purposes. The repos will have to be moved to somewhere else anyway before this is merged.

notspiff commented:
Got that, just so it isnt forgotten.
18. aug. 2015 10:58 skrev "Sascha Montellese" <notifications@github.com>:

> Yeah, this is just for testing purposes. The repos will have to be moved
> to somewhere else anyway before this is merged.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/7774#issuecomment-132125074>.
>

Montellese commented:
@FernetMenta @hudokkow @xhaggi @Jalle19: Does this complicate PVR addon development for you because the addon definition file isn't part of the kodi repository anymore? Or are you using `ADDON_SRC_PREFIX` to reference local PVR addon repository checkouts?

hudokkow commented:
I usually use <code>ADDON_SRC_PREFIX</code>
Don't know about the others.

Jalle19 commented:
@Montellese I mostly use Windows and there I just edit `pvr.hts.txt` to point at `file://path/to/source` and run your environment preparation script.

Montellese commented:
@Jalle19: with this change you'd first have to run the new `bootstrap-addons.bat`, then modify `pvr.hts.txt` and then run the environment preparation script.

Jalle19 commented:
Okay, fine by me as long as it's documented somewhere.

xhaggi commented:
Okay for me too

Montellese commented:
To get this ready for merging in the next merge window I'll do the following:
- [ ] strip down `bootstrap-addons.bat` to the parts that already work. The rest isn't needed just yet.
- [ ] write the `README` for the new binary addons bootstrap buildsystem
- [ ] fork the binary addon meta repositories into the `xbmc` account. @MartijnKaijser should these repositories be in the `xbmc` account or should the PVR meta repository be in the `kodi-pvr` account?
- [ ] create a `Jarvis` branch in the meta repositories and update the default meta repository definitions to point to that branch instead of `master`

hudokkow commented:
I think we should keep it under xbmc account. We're starting to get a lot of fragmentation (xbmc, kodi-pvr, kodi-adsp, notspiff, etc)

Montellese commented:
Moved the meta repository from my account to the xbmc account at https://github.com/xbmc/repo-binary-addons and adjusted the default repository that is bootstrapped. Simply executing the binary addons buildsystem will automatically bootstrap that repository on master branch (which points to the master branch of every binary addon repository.
On win32 it is also possible to manually execute `tools/buildsteps/win32/bootstrap-addons.bat` to achieve the same result. Would be nice if we could get something similar for the other platforms (for proper jenkins integration). I'll look into it if I got some time.

Montellese commented:
@wsnipex: Would you like to look into integrating the bootstrapping buildsystem into depends before this is merged or should we merge this as is? Right now it's not really necessary to be able to execute the bootstrapping step manually as it is automatically taken care of by the binary addons buildsystem.

wsnipex commented:
I'd leave it as is. Unified deps should use defaults imho.

Montellese commented:
@wsnipex: yes for normal builds but if we want to build specific versions with jenkins (for PRs and new addon releases) we need to be able to bootstrap a specific version.

wsnipex commented:
ok. I'd still rather do it later when needed instead of delaying this PR.

Montellese commented:
Sure, I'm totally fine with that.

Montellese commented:
jenkins build this please

Montellese commented:
jenkins build this please

MartijnKaijser commented:
jenkins build this please

hudokkow commented:
weird. jenkins failed after 0.41 sec

jenkins build this please

MartijnKaijser commented:
jenkins build this please

MartijnKaijser commented:
only failed for linux.
try again to be sure jenkins build this please

MartijnKaijser commented:
all green
